File: c++/611. Valid Triangle Number.cpp
class Solution {
public:
    int triangleNumber(vector<int>& nums) {
        int res = 0;
        sort(nums.begin(), nums.end());
        for (int i = nums.size()-1; i>=2; i--) {
            int l = 0, r = i-1;
            while (l < r) {
                if (nums[l] + nums[r] > nums[i]) {
                    res += r-l;  // 有效数目是l到r-1
                    r--;
                } else 
                    l++;
            }
        }
        return res;
    }
};

File: go/src/leetcode/611. Valid Triangle Number/611. Valid Triangle Number.go
package main

import (
	"fmt"
	"sort"
)

func triangleNumber(nums []int) int {
	sort.Ints(nums)
	n := len(nums)
	var count int
	for i := n-1; i >= 0; i-- {
		l, r := 0, i-1
		for l < r {
			if nums[l] + nums[r] > nums[i] {
				count += r - l
				r--
			} else {
				l++
			}
		}
	}
	fmt.Println(count)
	return count
}

func main() {
	triangleNumber([]int{2,2,3,4})
	triangleNumber([]int{4,2,3,4})
}
